src/pfb/importers/tsv.py

```diff
@@ -76,27 +76,52 @@ def _from_tsv(metadata, path, program, project):
             tsv_data = [tsv_data]
         for tsv_record in tsv_data:
             for k, v in tsv_record.items():
-                tsv_record[k] = convert_types(v)
+                field_type = get_type_from_schema(schema, node_name, k)
+                tsv_record[k] = convert_types(v, field_type)
             record = _convert_tsv(node_name, tsv_record, program, project, link_dests)
             yield record
 
 
-def convert_types(val):
-    if val is None or val.strip() == "":
+def convert_types(val, field_type):
+    if field_type == "string" or field_type == "enum":
+        if val is None or val.strip() == "":
+            return None
+        return str(val)
+    elif field_type == "float":
+        return float(val)
+    elif field_type == "integer" or field_type == "long":
+        return int(val)
+    elif field_type == "boolean":
+        if val.lower() == "false":
+            return False
+        if val.lower() == "true":
+            return True
+    else:
+        # finally if the type doesn't match any case then we return the supplied value
+        return val
+
+def get_type_from_schema(schema, node, field):
+    nodes = None
+    for n in schema:
+        if n["name"] == node:
+            nodes = n
+            break
+    if nodes == None:
         return None
-    if val.lower() == "false":
-        return False
-    if val.lower() == "true":
-        return True
-    try:
-        v = float(val)
-        # for fields that require type long
-        if int(v) == v:
-            v = int(v)
-        return v
-    except ValueError:
-        pass
-    return val
+
+    field_type = None
+    for f in nodes["fields"]:
+        if f["name"] == field:
+            if f["type"][1] == "null":
```

Comment:
it would be nice to set `x, y = f["type"]` for clarity, for readers who are not very familiar with the format - i don't know what `f["type"][0]` and `f["type"][1]` represent